packages/forms/src/model/abstract_model.ts

```diff
@@ -1336,20 +1319,182 @@ export abstract class AbstractControl<TValue = any, TRawValue extends TValue = T
+    // If parent has been marked artificially dirty we don't want to re-calculate the
+    // parent's dirtiness based on the children.
+    const skipPristineCheck = this._parentMarkedDirty(opts.onlySelf);
+
+    (this as Writable<this>).status = DISABLED;
+    (this as Writable<this>).errors = null;
+    this._forEachChild((control: AbstractControl) => {
+      control.disable({...opts, onlySelf: true});
+    });
+    this._updateValue();
+
+    if (opts.emitEvent !== false) {
+      (this.events as Subject<ControlEvent<TValue>>)
+          .next(new StatusChangeEvent(this.status, changedControl));
+      (this.events as Subject<ControlEvent<TValue>>)
+          .next(new ValueChangeEvent(this.value, changedControl));
+    }
+
+    this._updateAncestors({...opts, skipPristineCheck}, this);
+    this._onDisabledChange.forEach((changeFn) => changeFn(true));
+  }
+
+  /** @internal */
+  _markAsPending(opts: {onlySelf?: boolean, emitEvent?: boolean}, changedControl: AbstractControl):
+      void {
+    (this as Writable<this>).status = PENDING;
+
+    if (opts.emitEvent !== false) {
+      (this.events as Subject<ControlEvent<TValue>>)
+          .next(new StatusChangeEvent(this.status, changedControl));
+    }
+
+    if (this._parent && !opts.onlySelf) {
+      this._parent._markAsPending(opts, changedControl);
+    }
+  }
+
+  /** @internal */
+  _markAsDirty(opts: {onlySelf?: boolean}, changedControl: AbstractControl): void {
+    const changed = this.pristine === true;
+    (this as Writable<this>).pristine = false;
+
+    if (this._parent && !opts.onlySelf) {
+      this._parent._markAsDirty(opts, changedControl);
+    }
+
+    if (changed) {
+      (this.events as Subject<ControlEvent<TValue>>)
+          .next(new PristineChangeEvent(false, changedControl));
+    }
+  }
+
+  /** @internal */
+  _markAsPristine(opts: {onlySelf?: boolean}, changedControl: AbstractControl): void {
+    const changed = this.pristine === false;
+    (this as Writable<this>).pristine = true;
+    this._pendingDirty = false;
+
+    this._forEachChild((control: AbstractControl) => {
+      control.markAsPristine({onlySelf: true});
+    });
+
+    if (this._parent && !opts.onlySelf) {
+      this._parent._updatePristine(opts, this);
+    }
+
+    if (changed) {
+      (this.events as Subject<ControlEvent<TValue>>)
+          .next(new PristineChangeEvent(true, changedControl));
+    }
+  }
+
+  /** @internal */
+  _updatePristine(opts: {onlySelf?: boolean}, changedControl: AbstractControl): void {
+    const newPristine = !this._anyControlsDirty();
+    const changed = this.pristine !== newPristine;
     (this as Writable<this>).pristine = !this._anyControlsDirty();
 
     if (this._parent && !opts.onlySelf) {
-      this._parent._updatePristine(opts);
+      this._parent._updatePristine(opts, changedControl);
+    }
+
+    if (changed) {
+      (this.events as Subject<ControlEvent<TValue>>)
+          .next(new PristineChangeEvent(this.pristine, changedControl));
```

Comment:
This code is a bit repetitive. I think we can create a protected `emit` function on the `AbstractControl` class and use it in other places like this:
```suggestion
      this.emit(new PristineChangeEvent(this.pristine, changedControl));
```